llvm/lib/Target/ARM/ARMConstantIslandPass.cpp

```diff
@@ -1324,7 +1324,8 @@ bool ARMConstantIslands::findAvailableWater(CPUser &U, unsigned UserOffset,
   MachineBasicBlock *UserBB = U.MI->getParent();
   BBInfoVector &BBInfo = BBUtils->getBBInfo();
   const Align CPEAlign = getCPEAlign(U.CPEMI);
-  unsigned MinNoSplitDisp = BBInfo[UserBB->getNumber()].postOffset(CPEAlign);
+  unsigned MinNoSplitDisp =
+      BBInfo[UserBB->getNumber()].postOffset(CPEAlign) - UserOffset;
   if (CloserWater && MinNoSplitDisp > U.getMaxDisp() / 2)
     return false;
```

Comment:
I understand that this change means that `MinNoSplitDisp` is now the exact offset between the user of the constant-pool entry and the place we're considering putting the entry itself. But why are we then splitting the block if that's more than _half_ of `U.getMaxDisp()`? I would expect that if the distance we have to reach is actually _larger_ than `U.getMaxDisp()`, that's when we need to split something.

But I don't understand why the previous code was comparing against `U.getMaxDisp() / 2` either. Do you?

Reply:
I also have the same doubt actually. The check `MinNoSplitDisp > U.getMaxDisp() / 2` is only relevant when `CloserWater` is true which means the pass has not converged after 15 iterations with the default `arm-constant-island-max-iteration` being 30. I guess the original intention was that, given that the pass has not converged with this many iterations (in practice, it rarely seems to take more than 3 iterations to converge), maybe it should change its strategy to be more aggressive now (therefore comparing against `U.getMaxDisp() / 2` instead of `U.getMaxDisp()`)?